src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -312,11 +317,13 @@ internal override TypeWithAnnotations IteratorElementTypeWithAnnotations
             }
             set
             {
-                Debug.Assert(_lazyIteratorElementType == null || TypeSymbol.Equals(_lazyIteratorElementType.Value.Type, value.Type, TypeCompareKind.ConsiderEverything2));
-                Interlocked.CompareExchange(ref _lazyIteratorElementType, new TypeWithAnnotations.Boxed(value), null);
+                Debug.Assert(_lazyIteratorElementType == TypeWithAnnotations.Boxed.Sentinel || TypeSymbol.Equals(_lazyIteratorElementType.Value.Type, value.Type, TypeCompareKind.ConsiderEverything2));
+                Interlocked.CompareExchange(ref _lazyIteratorElementType, new TypeWithAnnotations.Boxed(value), TypeWithAnnotations.Boxed.Sentinel);
             }
         }
 
+        internal override bool IsIterator => _lazyIteratorElementType is object;
```

Comment:
>_lazyIteratorElementType [](start = 45, length = 24)

`IteratorElementTypeWithAnnotations`

Reply:
> IteratorElementTypeWithAnnotations.HasType

That would  be incorrect, we want a null check here. Besides, ```TypeWithAnnotations.Boxed``` doesn't have HasType property.

---
In reply to: [384804843](https://github.com/dotnet/roslyn/pull/41926#discussion_r384804843) [](ancestors = 384804843)

Reply:
This property returns `true` until `IteratorElementTypeWithAnnotations.set` is called which is confusing.

---
In reply to: [384809339](https://github.com/dotnet/roslyn/pull/41926#discussion_r384809339) [](ancestors = 384809339,384804843)

Reply:
> That would  be incorrect, we want a null check here. 

Even if for whatever reason we cannot calculate a type, or have it ```default(TypeWithAnnotations)```, the method is still an iterator. That is a syntax check. 

---
In reply to: [384809339](https://github.com/dotnet/roslyn/pull/41926#discussion_r384809339) [](ancestors = 384809339,384804843)

Reply:
> This property returns true until IteratorElementTypeWithAnnotations.set is called which is confusing.

After constructor is done, ```_lazyIteratorElementType``` never changes its null state. It can change from a non-null value to a different non-null value, but never from null to non-null or vise versa.

---
In reply to: [384811908](https://github.com/dotnet/roslyn/pull/41926#discussion_r384811908) [](ancestors = 384811908,384809339,384804843)